sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -145,14 +159,56 @@ object Hash {
       add(longBuffer)
     }
 
+    private def add(fieldNumber: Int, v: Long): this.type =
+      if (v != 0) add(fieldNumber).add(v) else this
+
+    final def add(v: data.Numeric): this.type = {
+      val a = v.unscaledValue().toByteArray
+      add(a.length).add(a)
+    }
+
+    private def add(fieldNumber: Int, v: data.Numeric): this.type =
+      if (v.signum() != 0) add(fieldNumber).add(v) else this
+
     final def iterateOver[T, U](a: ImmArray[T])(f: (this.type, T) => this.type): this.type =
       a.foldLeft[this.type](add(a.length))(f)
 
+    private def iterateOver[T, U](fieldNumber: Int, a: ImmArray[T])(
+        f: (this.type, T) => this.type
+    ): this.type =
+      if (a.nonEmpty)
+        a.foldLeft[this.type](add(fieldNumber).add(a.length))(f)
+      else
+        this
+
     final def iterateOver[T](a: Iterator[T], size: Int)(f: (this.type, T) => this.type): this.type =
       a.foldLeft[this.type](add(size))(f)
 
-    final def iterateOver[T](a: Option[T])(f: (this.type, T) => this.type): this.type =
-      a.fold[this.type](add(0))(f(add(1), _))
+    final def iterateOver[T](fieldNumber: Int, a: Iterator[T], size: Int)(
+        f: (this.type, T) => this.type
+    ): this.type =
+      if (a.nonEmpty)
+        a.foldLeft[this.type](add(fieldNumber).add(size))(f)
+      else
+        this
+
+    final def iterateOver[T](v: Option[T])(f: (this.type, T) => this.type): this.type =
+      v match {
+        case Some(v) =>
+          f(add(1), v)
+        case None =>
+          add(0)
```

Comment:
You're using four bytes to distinguish the `None` constructor from `Some`. How about using `Byte`s instead of `Int`s?
```suggestion
          f(add(1B), v)
        case None =>
          add(0B)
```

Reply:
Yes. Unfortunatly, I introduce this error. 
We have to stick to it :(
I will remove completely the overloading to make it more obvious.

Reply:
354e725736dc66801cfc90df43e8a638644b254e